File: apps/api/routers/compare.py
"""
Activity Comparison API Router

Enables powerful workout comparisons that Garmin/Strava don't offer:
- Compare all tempo runs
- Compare hot vs cool days
- Compare this month vs last month
- See trends within workout types
"""

from fastapi import APIRouter, Depends, HTTPException, Query, Body
from sqlalchemy.orm import Session
from typing import Optional, List
from datetime import datetime, timedelta
from uuid import UUID
from pydantic import BaseModel, Field

from core.database import get_db
from core.auth import get_current_athlete
from models import Athlete, Activity
from services.activity_comparison import ActivityComparisonService
from services.workout_classifier import WorkoutClassifierService

router = APIRouter(prefix="/v1/compare", tags=["Comparison"])


class IndividualCompareRequest(BaseModel):
    """Request body for individual activity comparison"""
    activity_ids: List[UUID] = Field(
        ..., 
        min_length=2, 
        max_length=10,
        description="List of 2-10 activity IDs to compare"
    )


@router.post("/individual")
async def compare_individual_activities(
    request: IndividualCompareRequest,
    athlete: Athlete = Depends(get_current_athlete),
    db: Session = Depends(get_db),
):
    """
    Compare 2-10 specific activities side-by-side.
    
    This is the marquee feature - select specific runs and see them
    compared with metrics table, best/worst highlighting, and insights.
    
    Returns:
    - All activity details in a comparison table
    - Best performer for each metric
    - Plain-language insights
    """
    service = ActivityComparisonService(db)
    
    try:
        result = service.compare_individual(
            athlete_id=athlete.id,
            activity_ids=request.activity_ids,
        )
        return result.to_dict()
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/workout-types")
async def get_workout_type_summary(
    athlete: Athlete = Depends(get_current_athlete),
    db: Session = Depends(get_db),
):
    """
    Get a summary of activities by workout type.
    Shows how many of each type you have for comparison.
    """
    service = ActivityComparisonService(db)
    summary = service.get_workout_type_summary(athlete.id)
    
    return {
        "workout_types": summary,
        "total": sum(summary.values()),
    }


@router.get("/by-type/{workout_type}")
async def compare_by_workout_type(
    workout_type: str,
    days: int = Query(default=180, ge=7, le=730),
    athlete: Athlete = Depends(get_current_athlete),
    db: Session = Depends(get_db),
):
    """
    Compare all activities of a specific workout type.
    
    Returns:
    - Trend analysis (improving/declining/stable)
    - Best and worst performances
    - All activities for detailed view
    """
    service = ActivityComparisonService(db)
    result = service.compare_by_workout_type(
        athlete_id=athlete.id,
        workout_type=workout_type,
        days=days,
    )
    
    return result.to_dict()


@router.get("/by-conditions")
async def compare_by_conditions(
    workout_type: Optional[str] = None,
    temp_min: Optional[float] = Query(default=None, description="Min temperature in °F"),
    temp_max: Optional[float] = Query(default=None, description="Max temperature in °F"),
    days: int = Query(default=365, ge=30, le=730),
    athlete: Athlete = Depends(get_current_athlete),
    db: Session = Depends(get_db),
):
    """
    Compare activities filtered by conditions.
    
    Example: Compare tempo runs in hot weather (>80°F) vs cool (<60°F)
    """
    service = ActivityComparisonService(db)
    result = service.compare_by_conditions(
        athlete_id=athlete.id,
        workout_type=workout_type,
        temp_min=temp_min,
        temp_max=temp_max,
        days=days,
    )
    
    return result.to_dict()


@router.get("/activity/{activity_id}")
async def get_activity_with_comparison(
    activity_id: UUID,
    athlete: Athlete = Depends(get_current_athlete),
    db: Session = Depends(get_db),
):
    """
    Get an activity with comparison to similar workouts.
    
    Returns:
    - Activity details with splits
    - Comparison to similar workout types
    - Percentile ranking
    """
    service = ActivityComparisonService(db)
    result = service.get_activity_with_comparison(
        activity_id=activity_id,
        athlete_id=athlete.id,
    )
    
    if not result:
        raise HTTPException(status_code=404, detail="Activity not found")
    
    return result


@router.post("/classify-all")
async def classify_all_activities(
    athlete: Athlete = Depends(get_current_athlete),
    db: Session = Depends(get_db),
):
    """
    Classify all activities that don't have a workout type yet.
    This is a one-time operation to backfill existing activities.
    """
    classifier = WorkoutClassifierService(db)
    
    # Get unclassified activities
    activities = db.query(Activity).filter(
        Activity.athlete_id == athlete.id,
        Activity.workout_type.is_(None),
        Activity.distance_m >= 1000,  # Only classify meaningful runs
    ).all()
    
    classified_count = 0
    for activity in activities:
        try:
            classification = classifier.classify_activity(activity)
            activity.workout_type = classification.workout_type.value
            activity.workout_zone = classification.workout_zone.value
            activity.workout_confidence = classification.confidence
            activity.intensity_score = classification.intensity_score
            classified_count += 1
        except Exception as e:
            # Skip activities that fail classification
            continue
    
    db.commit()
    
    return {
        "classified": classified_count,
        "total_unclassified_before": len(activities),
        "message": f"Classified {classified_count} activities"
    }

File: apps/web/lib/api/services/compare.ts
/**
 * Activity Comparison API Service
 * 
 * Enables comparison of workouts by type, conditions, time periods.
 * Key differentiator from Garmin/Strava.
 */

import { apiClient } from '../client';

export interface ChartSplitData {
  split_number: number;
  distance_m: number;
  elapsed_time_s: number;
  pace_per_km: number | null;
  avg_hr: number | null;
  cumulative_distance_m: number;
}

export interface ActivitySummary {
  id: string;
  date: string;
  name: string;
  workout_type: string | null;
  distance_m: number;
  distance_km: number;
  duration_s: number;
  avg_hr: number | null;
  pace_per_km: number | null;
  pace_formatted: string | null;
  efficiency: number | null;
  intensity_score: number | null;
  elevation_gain: number | null;
  temperature_f: number | null;
  splits: ChartSplitData[];
}

export interface ComparisonResult {
  workout_type: string;
  total_activities: number;
  date_range_start: string | null;
  date_range_end: string | null;
  avg_pace_per_km: number | null;
  avg_pace_formatted: string | null;
  avg_efficiency: number | null;
  avg_hr: number | null;
  total_distance_km: number;
  efficiency_trend: 'improving' | 'declining' | 'stable' | null;
  efficiency_change_pct: number | null;
  pace_trend: 'improving' | 'declining' | 'stable' | null;
  pace_change_pct: number | null;
  best_activity: ActivitySummary | null;
  worst_activity: ActivitySummary | null;
  most_recent: ActivitySummary | null;
  activities: ActivitySummary[];
}

export interface WorkoutTypeSummary {
  workout_types: Record<string, number>;
  total: number;
}

export interface ActivityWithComparison {
  id: string;
  date: string;
  workout_type: string | null;
  workout_zone: string | null;
  distance_m: number;
  duration_s: number;
  avg_hr: number | null;
  max_hr: number | null;
  elevation_gain: number | null;
  pace_per_km: number | null;
  efficiency: number | null;
  intensity_score: number | null;
  temperature_f: number | null;
  splits: SplitData[];
  comparison: {
    similar_count: number;
    percentile_vs_similar: number | null;
    avg_similar_efficiency: number | null;
    similar_activities: Array<{
      id: string;
      date: string;
      distance_m: number;
      pace_per_km: number | null;
      avg_hr: number | null;
      efficiency: number | null;
    }>;
  } | null;
}

export interface SplitData {
  split_number: number;
  distance: number | null;
  elapsed_time: number | null;
  moving_time: number | null;
  average_heartrate: number | null;
  max_heartrate: number | null;
  average_cadence: number | null;
  gap_seconds_per_mile: number | null;
}

export interface ClassifyResult {
  classified: number;
  total_unclassified_before: number;
  message: string;
}

export interface IndividualComparisonResult {
  activities: ActivitySummary[];
  comparison_table: Record<string, (string | number | null)[]>;
  best_by_metric: Record<string, string>;
  insights: string[];
  count: number;
}

export const compareService = {
  /**
   * Get summary of activities by workout type
   */
  async getWorkoutTypeSummary(): Promise<WorkoutTypeSummary> {
    return apiClient.get<WorkoutTypeSummary>('/v1/compare/workout-types');
  },

  /**
   * Compare activities by workout type
   */
  async compareByType(workoutType: string, days: number = 180): Promise<ComparisonResult> {
    return apiClient.get<ComparisonResult>(`/v1/compare/by-type/${workoutType}?days=${days}`);
  },

  /**
   * Compare activities by conditions
   */
  async compareByConditions(params: {
    workout_type?: string;
    temp_min?: number;
    temp_max?: number;
    days?: number;
  }): Promise<ComparisonResult> {
    const searchParams = new URLSearchParams();
    if (params.workout_type) searchParams.set('workout_type', params.workout_type);
    if (params.temp_min !== undefined) searchParams.set('temp_min', params.temp_min.toString());
    if (params.temp_max !== undefined) searchParams.set('temp_max', params.temp_max.toString());
    if (params.days) searchParams.set('days', params.days.toString());
    
    return apiClient.get<ComparisonResult>(`/v1/compare/by-conditions?${searchParams}`);
  },

  /**
   * Get activity with comparison to similar workouts
   */
  async getActivityWithComparison(activityId: string): Promise<ActivityWithComparison> {
    return apiClient.get<ActivityWithComparison>(`/v1/compare/activity/${activityId}`);
  },

  /**
   * Classify all unclassified activities
   */
  async classifyAll(): Promise<ClassifyResult> {
    return apiClient.post<ClassifyResult>('/v1/compare/classify-all');
  },

  /**
   * Compare 2-10 specific activities side-by-side
   * This is the marquee feature
   */
  async compareIndividual(activityIds: string[]): Promise<IndividualComparisonResult> {
    return apiClient.post<IndividualComparisonResult>('/v1/compare/individual', {
      activity_ids: activityIds,
    });
  },
};
